File: icm.py
import numpy as np
import itertools

stacks = [1000, 200, 100, 390, 72, 50]

def calc_prob_places(stacks : list):
    # This matrix (row=i,column=j) represents prob that stack i comes in place j
    prob_matrix = np.zeros((len(stacks), len(stacks)))

    # The first row is probability that stack i comes in first place
    prob_matrix[:,0] = np.array(stacks)/np.sum(stacks)

    sum_stacks = sum(stacks)
    # Second column
    for i in range(len(stacks)):
        prob_matrix[i,1] = sum([stacks[j] * stacks[i]/((sum_stacks - stacks[j])*sum_stacks) for j in range(len(stacks)) if j!= i])
    
    # Third column
    for i in range(len(stacks)):
        prob_matrix[i, 2] = sum([
            stacks[j] * stacks[k] * stacks[i] / ((sum_stacks - stacks[j] - stacks[k]) * (sum_stacks - stacks[j]) * sum_stacks)
            for j in range(len(stacks)) for k in range(len(stacks))
            if j != i and k != j and k != i
        ])

    # Fourth column
    if len(stacks) > 3:
        for i in range(len(stacks)):
            prob_matrix[i,3] = sum([
                stacks[i] * stacks[j] * stacks[k] * stacks[l] / (
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l] ) * 
                    (sum_stacks - stacks[j] - stacks[k]) * 
                    (sum_stacks - stacks[j]) 
                    * sum_stacks
                )
                for j in range(len(stacks)) for k in range(len(stacks)) for l in range(len(stacks))
                if j != i and k != j and k != i and l != k and l != j and l != i
            ])
    
    # Fifth column
    if len(stacks) > 4:
        for i in range(len(stacks)):
            prob_matrix[i,4] = sum([
                stacks[i] * stacks[j] * stacks[k] * stacks[l] * stacks[m] / (
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l] - stacks[m]) * 
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l]) * 
                    (sum_stacks - stacks[j] - stacks[k]) * 
                    (sum_stacks - stacks[j]) 
                    * sum_stacks
                )
                for j in range(len(stacks)) for k in range(len(stacks)) for l in range(len(stacks)) for m in range(len(stacks))
                if j != i and k != j and k != i and l != k and l != j and l != i and m != l and m != k and m != j and m != i
            ])
    
    # Sixth column
    if len(stacks) > 5:
        for i in range(len(stacks)):
            prob_matrix[i,5] = sum([
                stacks[i] * stacks[j] * stacks[k] * stacks[l] * stacks[m] * stacks[n] / (
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l] - stacks[m] - stacks[n]) * 
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l] - stacks[m]) * 
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l]) * 
                    (sum_stacks - stacks[j] - stacks[k]) * 
                    (sum_stacks - stacks[j]) 
                    * sum_stacks
                )
                for j in range(len(stacks)) for k in range(len(stacks)) for l in range(len(stacks)) for m in range(len(stacks)) for n in range(len(stacks))
                if j != i and k != j and k != i and l != k and l != j and l != i and m != l and m != k and m != j and m != i and n != m and n != l and n != k and n != j and n != i
            ])
        
    # Seventh column
    if len(stacks) > 6:
        for i in range(len(stacks)):
            prob_matrix[i,6] = sum([
                stacks[i] * stacks[j] * stacks[k] * stacks[l] * stacks[m] * stacks[n] * stacks[o] / (
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l] - stacks[m] - stacks[n] - stacks[o]) * 
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l] - stacks[m] - stacks[n]) * 
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l] - stacks[m]) * 
                    (sum_stacks - stacks[j] - stacks[k] - stacks[l]) * 
                    (sum_stacks - stacks[j] - stacks[k]) * 
                    (sum_stacks - stacks[j]) 
                    * sum_stacks
                )
                for j in range(len(stacks)) for k in range(len(stacks)) for l in range(len(stacks)) for m in range(len(stacks)) for n in range(len(stacks)) for o in range(len(stacks))
                if j != i and k != j and k != i and l != k and l != j and l != i and m != l and m != k and m != j and m != i and n != m and n != l and n != k and n != j and n != i and o != n and o != m and o != l and o != k and o != j and o != i
            ])
    return prob_matrix

def calc_prob(current, remaining, total, depth=6):
    if len(remaining) == 0:
        prob = 1
        for i, stack in enumerate(current):
            if i == 0 :
                prob *= stack / total
            elif i < depth:
                total -= stack
                prob *= stack / total
        return prob
    else:
        return sum(calc_prob(current + [stack], [r for r in remaining if r != stack], total, depth) for stack in remaining)

def calc_prob_matrix(stacks):
    prob_matrix = np.zeros((len(stacks), len(stacks)))
    total = sum(stacks)

    for j in range(prob_matrix.shape[1]):
        for i in range(len(stacks)):
            prob_matrix[i, j] = calc_prob([stacks[i]], [stack for k, stack in enumerate(stacks) if k != i], total, depth=j+1)
    return prob_matrix

def make_perms(stacks):
    all_perms = list(list(a) for a in itertools.permutations(stacks))
    split = len(all_perms)//len(stacks)
    stack_dict = {i : all_perms[i*split:(i+1)*split] for i in range(len(stacks))}
    return stack_dict

def calc_prob(current, total):
    prob = 1
    for i, stack in enumerate(current):
        if i == 0 :
            prob *= stack / total
        else:
            total -= stack
            prob *= stack / total
    return prob

def cut_perms(perms, n, stack_length=6):
    repeat_len = np.math.factorial(stack_length - n - 1)
    cut_size = len(perms)//repeat_len
    cut_list = [repeat_len*i for i in range(cut_size)]
    return [perms[cut_index][:n+1] for cut_index in cut_list]


def calc_prob_matrix(stacks):
    prob_matrix = np.zeros((len(stacks), len(stacks)))
    total = sum(stacks)
    stack_orderings = make_perms(stacks)

    for j in range(prob_matrix.shape[1]):
        for i in range(prob_matrix.shape[0]):
            cutted_perms = cut_perms(stack_orderings[i], j, len(stacks))
            prob_matrix[i, j] = sum(calc_prob(cutted_perms[l], total) for l in range(len(cutted_perms)))
    return prob_matrix


c = calc_prob_matrix(stacks)
print(c)

File: rust_version/src/icm.rs
use itertools::Itertools;
use factorial::Factorial;
use ndarray::Array1;
use ndarray::Array2;
use ndarray::arr1;

pub struct ICMSolver {
    pub stacks: Vec<f64>,
    pub prob_matrix: Array2<f64>,
    pub payouts: Vec<f64>,
    pub icm_values: Vec<f64>,
}

impl ICMSolver {
    pub fn new(stacks: Vec<f64>, payouts: Vec<f64>) -> Self {
        let prob_matrix = Self::calc_prob_matrix(&stacks);
        let icm_values = prob_matrix.dot(&arr1(&payouts)).to_vec();
        Self { stacks, prob_matrix, payouts, icm_values}
    }

    fn calc_prob(current: &[f64], total: f64) -> f64 {
        let mut prob = 1.0;
        let mut total = total;
        for (i, &stack) in current.iter().enumerate() {
            if i == 0 {
                prob *= stack / total;
            } else {
                total -= stack;
                prob *= stack / total;
            }
        }
        prob
    }

    fn cut_perms(perms: &[Vec<f64>], n: usize, stack_length: usize) -> Vec<Vec<f64>> {
        let repeat_size = (stack_length - n - 1).factorial();
        let cut_size: usize = perms.len() / repeat_size;
        let cut_list: Vec<usize> = (0..cut_size).map(|i| repeat_size * i).collect();    
        cut_list.iter().map(|&cut_index| perms[cut_index][..=n].to_vec()).collect()
    }

    fn make_stack_orderings(stacks: Vec<f64>) -> Vec<Vec<Vec<f64>>> {
        let stack_length = stacks.len();
        let stack_orderings: Vec<_> = stacks.iter().cloned().permutations(stack_length).collect();
        let split = stack_orderings.len() as u32 / stacks.len() as u32;
        let mut ordered_stack_orderings = vec![];
        for i in 0..stacks.len() {
            let start = i * split as usize;
            let end = (i + 1) * split as usize;
            ordered_stack_orderings.push(stack_orderings[start..end].to_vec());
        }
        ordered_stack_orderings
    }

    fn calc_prob_matrix(stacks: &Vec<f64>) -> Array2<f64> {
        let total = stacks.iter().sum::<f64>();
        let stack_orderings: Vec<_> = Self::make_stack_orderings(stacks.clone());
        let mut prob_matrix = Array2::zeros((stacks.len(), stacks.len()));
    
        for j in 0..prob_matrix.ncols() {
            for i in 0..prob_matrix.nrows() {
                let cutted_perms = Self::cut_perms(&stack_orderings[i], j, stacks.len());
                prob_matrix[[i, j]] = cutted_perms.iter().map(|perm| Self::calc_prob(perm, total)).sum::<f64>();
            }
        }
        prob_matrix
    }
}

pub struct RiskPremium {
    pub stacks: Vec<f64>,
    pub payouts: Vec<f64>,
    pub neutral_icm: Array1<f64>,
    pub win_matrix: Array2<f64>,
    pub lose_matrix: Array2<f64>,
    pub rp_matrix: Array2<f64>,
}

impl RiskPremium {
    pub fn new(stacks: Vec<f64>, payouts: Vec<f64>) -> Self {
        let icm_solver = ICMSolver::new(stacks.clone(), payouts.clone());
        let neutral_icm = arr1(&icm_solver.icm_values);
        let mut rp = Self {
            stacks,
            payouts,
            neutral_icm,
            win_matrix: Array2::zeros((0, 0)), // temporary values
            lose_matrix: Array2::zeros((0, 0)), // temporary values
            rp_matrix: Array2::zeros((0, 0)), // temporary values
        };
        let (win_matrix, lose_matrix) = rp.calc_stackoff_matrix();
        rp.win_matrix = win_matrix;
        rp.lose_matrix = lose_matrix;
        rp.rp_matrix = rp.calc_rp_matrix();
        rp
    }

    fn calc_stackoff_matrix(&mut self) -> (Array2<f64>, Array2<f64>) {
        let len = self.stacks.len();
        let mut a = Array2::zeros((len, len));
        let mut b = Array2::zeros((len, len));
        for i in 0..len {
            for j in 0..len {
                let mut new_stacks = self.stacks.clone();
                if i == j {
                    a[[i, i]] = self.neutral_icm[i];
                } else {
                    let loser_stack = f64::max(0.000000001, new_stacks[j] - new_stacks[i]);
                    let winner_stack = f64::min(new_stacks[j] + new_stacks[i], 2.0 * new_stacks[i]);
                    new_stacks[i] = winner_stack;
                    new_stacks[j] = loser_stack;
                    let icm_solver = ICMSolver::new(new_stacks, self.payouts.clone());
                    a[[i, j]] = icm_solver.icm_values[i];
                    b[[j, i]] = icm_solver.icm_values[j];
                }
            }
        }
        (a, b)
    }

    fn calc_rp_matrix(&self) -> Array2<f64> {
        let neutral_matrix = (self.neutral_icm.clone() * Array2::<f64>::ones((self.stacks.len(), self.stacks.len()))).reversed_axes();
        let numerator = &neutral_matrix - &self.lose_matrix;
        let denominator = &self.win_matrix - &self.lose_matrix;
        (&numerator / &denominator - 0.5 * Array2::eye(self.stacks.len()) - 0.5 * Array2::ones((self.stacks.len(), self.stacks.len()))) * 100.0
    }
}